zrml/neo-swaps/src/lib.rs

```diff
@@ -979,4 +979,89 @@ mod pallet {
             Self::do_deploy_pool(who, market_id, amount, spot_prices, swap_fee)
         }
     }
+
+    impl<T: Config> Pallet<T> {
+        fn amount_including_fee_surplus(
+            amount: BalanceOf<T>,
+            fee_fractional: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let fee_divsisor = ZeitgeistBase::<BalanceOf<T>>::get()?
+                .checked_sub(&fee_fractional)
+                .ok_or(Error::<T>::Unexpected)?;
+            amount.bdiv(fee_divsisor)
```

Comment:
Why are we `amount / (1-fee_fractional)` here? Basically we are asking "What is the amount of which deducting `fee_fractional` results in `amount`" instead of what I would've expected "What is the amount to pay when charging `fee_factor` on top of `amount`?" which would be `amount * fee_fractional`.

Reply:
https://mathsathome.com/reverse-percentages/

We have `amount_without_fees = calculate_buy_amount_until(...)`. The AMM buy tx charges fees **before**  calculating the amount_out using amount_in_minus_fees [here](https://github.com/zeitgeistpm/zeitgeist/blob/be3efc609c3f0e97ef6970ad101de2904374384e/zrml/neo-swaps/src/lib.rs#L549-L560). So yes, intuitively you think, just add fees on top of the buy amount by: `buy_amount = amount + amount * fee`. Like this:

```rust
        fn calculate_buy_amount_until(
            market_id: Self::MarketId,
            asset: Self::Asset,
            until: Self::Balance,
        ) -> Result<Self::Balance, DispatchError> {
            let pool = Pools::<T>::get(market_id).ok_or(Error::<T>::PoolNotFound)?;
            let buy_amount = pool.calculate_buy_amount_until(asset, until)?;
            let external_fee_percentage = T::ExternalFees::fee_percentage(market_id);
            let external_fee_amount = external_fee_percentage.mul_floor(buy_amount);
            let swap_fee_amount = pool.swap_fee.bmul(buy_amount)?;
            let total_fee_amount = swap_fee_amount.saturating_add(external_fee_amount);
            let remaining =
                buy_amount.checked_sub(&total_fee_amount).ok_or(Error::<T>::Unexpected)?;
            Ok(remaining)
        }
```

But this `buy_amount` adds fees on top of `amount`, which is already operating on the reduced `amount` basis. In order to get the basis of the `amount` before the fees were charged, you need `amount / (1-fee)`.

Example: If you charge 10% fees on 100, you get a remainder of 90, which is the basis of the reduced `amount` now. Now, if you add 10% on the basis of the already reduced `amount` you would calculate: `90 * 1.1 = 99`, so not 100. In order to get from 90 to 100, we need to calculate `90 / (1 - 0.1) = 100`.

In our case, the `calculate_buy_amount_until` returns the amount, which is already on the reduced basis, because the AMM will first charge the fee percentage on the amount with the bumped basis (in the example the 100) and then calculate the amount_out using the `amount_minus_fees`. If the AMM buy tx had first used the `amount_in` for their `amount_out` calculation (not `amount_minus_fees`) and then charged the fees, our `calculate_buy_amount_until` would have already yielded the correct result. That's why `calculate_sell_amount_until` disregards fees (does not add them).

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1275/commits/9731b37fd3d4101cc6176910abffc1ffeaf7b033

Reply:
I see, didn't know it was already reduced.